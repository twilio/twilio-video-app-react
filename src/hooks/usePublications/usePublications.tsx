import { useEffect, useState } from 'react';
import { LocalTrackPublication, Participant, RemoteTrackPublication } from 'twilio-video';

type TrackPublication = LocalTrackPublication | RemoteTrackPublication;

export default function usePublications(participant: Participant) {
  const [publications, setPublications] = useState<TrackPublication[]>([]);

  useEffect(() => {
    // Reset the publications when the 'participant' variable changes.
    setPublications(Array.from(participant.tracks.values()) as TrackPublication[]);

    const publicationAdded = (publication: TrackPublication) =>
      setPublications(prevPublications => [...prevPublications, publication]);
    const publicationRemoved = (publication: TrackPublication) =>
      setPublications(prevPublications => prevPublications.filter(p => p !== publication));

    participant.on('trackPublished', publicationAdded);
    participant.on('trackUnpublished', publicationRemoved);
    return () => {
      participant.off('trackPublished', publicationAdded);
      participant.off('trackUnpublished', publicationRemoved);
    };
  }, [participant]);

  return publications;
}
