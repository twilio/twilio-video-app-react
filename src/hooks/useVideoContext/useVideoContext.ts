import { useContext } from 'react';
import { VideoContext } from '../../components/VideoProvider';

export default function useVideoContext() {
  const context = useContext(VideoContext);
  if (!context) {
    throw new Error('useVideoContext must be used within a VideoProvider');
  }
  return context;
}
