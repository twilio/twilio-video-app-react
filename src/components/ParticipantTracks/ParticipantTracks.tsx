import React from 'react';
import { LocalParticipant, RemoteParticipant } from 'twilio-video';
import Publication from '../Publication/Publication';
import usePublications from '../../hooks/usePublications/usePublications';
import { useVideoContext } from '../../hooks/context';

interface ParticipantTracksProps {
  participant: LocalParticipant | RemoteParticipant;
}

export default function ParticipantTracks({
  participant,
}: ParticipantTracksProps) {
  const { room } = useVideoContext();
  const publications = usePublications(participant);
  const isLocal = participant === room.localParticipant;

  return (
    <>
      {publications.map(publication => (
        <Publication
          key={publication.trackSid}
          publication={publication}
          participant={participant}
          isLocal={isLocal}
        />
      ))}
    </>
  );
}
