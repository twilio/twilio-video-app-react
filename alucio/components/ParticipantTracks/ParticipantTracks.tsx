import React from 'react';
import { Participant, Track } from 'twilio-video';
import { IVideoTrack } from 'src/packages/video/src/types';
import { Iffy } from '@alucio/lux-ui'
import Publication from '../../components/Publication/Publication';
import ParticipantPlaceholder from '../ParticipantInfo/ParticipantPlaceholder/ParticipantPlaceholder'
import usePublications from '../../../src/hooks/usePublications/usePublications';
import useVideoContext from '../../../src/hooks/useVideoContext/useVideoContext';
import useTrack from '../../../src/hooks/useTrack/useTrack'

interface ParticipantTracksProps {
  participant: Participant;
  disableAudio?: boolean;
  enableScreenShare?: boolean;
  videoPriority?: Track.Priority | null;
}

/*
 *  The object model for the Room object (found here: https://www.twilio.com/docs/video/migrating-1x-2x#object-model) shows
 *  that Participant objects have TrackPublications, and TrackPublication objects have Tracks.
 *
 *  The React components in this application follow the same pattern. This ParticipantTracks component renders Publications,
 *  and the Publication component renders Tracks.
 */

export default function ParticipantTracks({
  participant,
  disableAudio,
  enableScreenShare,
  videoPriority,
}: ParticipantTracksProps) {
  const { room } = useVideoContext();
  const publications = usePublications(participant);
  const isLocal = participant === room.localParticipant;

  let filteredPublications;

  if (enableScreenShare && publications.some(p => p.trackName.includes('screen'))) {
    filteredPublications = publications.filter(p => !p.trackName.includes('camera'));
  } else {
    filteredPublications = publications.filter(p => !p.trackName.includes('screen'));
  }

  const videoPublication = filteredPublications.find(p => p.trackName.includes('camera'))
  const videoTrack = useTrack(videoPublication) as IVideoTrack
  const showPlaceholder = !(
    videoTrack?.isEnabled
    // videoTrack?.isStarted is not apart of the React lifecycle so renders dependant on it are flaky
    // && videoTrack?.isStarted
  )

  return (
    <>
      {
        filteredPublications.map((publication, idx) => (
          <Publication
            key={`${publication.kind}-${idx}`}
            publication={publication}
            participant={participant}
            isLocal={isLocal}
            disableAudio={disableAudio}
            videoPriority={videoPriority}
          />
        ))
      }
      {/* Assume all other Publications contain no video track in a ready state */}
      <Iffy is={showPlaceholder}>
        <ParticipantPlaceholder />
      </Iffy>
    </>
  );
}
