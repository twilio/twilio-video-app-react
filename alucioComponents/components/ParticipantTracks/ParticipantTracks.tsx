import React from 'react';
import { Participant, Track } from 'twilio-video';
import Publication from '../../components/Publication/Publication';
import usePublications from '../../../src/hooks/usePublications/usePublications';
import useVideoContext from '../../../src/hooks/useVideoContext/useVideoContext';

interface ParticipantTracksProps {
  participant: Participant;
  disableAudio?: boolean;
  enableScreenShare?: boolean;
  videoPriority?: Track.Priority | null;
}

/*
 *  The object model for the Room object (found here: https://www.twilio.com/docs/video/migrating-1x-2x#object-model) shows
 *  that Participant objects have TrackPublications, and TrackPublication objects have Tracks.
 *
 *  The React components in this application follow the same pattern. This ParticipantTracks component renders Publications,
 *  and the Publication component renders Tracks.
 */

export default function ParticipantTracks({
  participant,
  disableAudio,
  enableScreenShare,
  videoPriority,
}: ParticipantTracksProps) {
  const { room } = useVideoContext();
  const publications = usePublications(participant);
  const isLocal = participant === room.localParticipant;

  let filteredPublications;

  if (enableScreenShare && publications.some(p => p.trackName.includes('screen'))) {
    filteredPublications = publications.filter(p => !p.trackName.includes('camera'));
  } else {
    filteredPublications = publications.filter(p => !p.trackName.includes('screen'));
  }

  return (
    <>
      {filteredPublications.map(publication => (
        <Publication
          key={publication.kind}
          publication={publication}
          participant={participant}
          isLocal={isLocal}
          disableAudio={disableAudio}
          videoPriority={videoPriority}
        />
      ))}
    </>
  );
}
