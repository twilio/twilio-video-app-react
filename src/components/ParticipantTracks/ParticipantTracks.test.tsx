import React from 'react';
import ParticipantTracks from './ParticipantTracks';
import usePublications from '../../hooks/usePublications/usePublications';
import { shallow } from 'enzyme';
import useVideoContext from '../../hooks/useVideoContext/useVideoContext';

jest.mock('../../hooks/usePublications/usePublications', () =>
  jest.fn(() => [
    { trackSid: 0, kind: 'video' },
    { trackSid: 1, kind: 'audio' },
  ])
);
jest.mock('../../hooks/useVideoContext/useVideoContext');

const mockUsePublications = usePublications as jest.Mock<any>;
const mockUseVideoContext = useVideoContext as jest.Mock<any>;

describe('the ParticipantTracks component', () => {
  it('should render an array of publications', () => {
    mockUseVideoContext.mockImplementation(() => ({ room: {} }));
    const wrapper = shallow(<ParticipantTracks participant={'mockParticipant' as any} />);
    expect(usePublications).toHaveBeenCalledWith('mockParticipant');
    expect(wrapper).toMatchSnapshot();
  });

  it('should render publications with "isLocal" set to true when the localParticipant is provided', () => {
    mockUseVideoContext.mockImplementation(() => ({
      room: { localParticipant: 'mockParticipant' },
    }));
    const wrapper = shallow(<ParticipantTracks participant={'mockParticipant' as any} />);
    expect(
      wrapper
        .find('Publication')
        .first()
        .prop('isLocal')
    ).toEqual(true);
  });

  it('should filter out any screen share publications', () => {
    mockUseVideoContext.mockImplementation(() => ({ room: {} }));
    mockUsePublications.mockImplementation(() => [
      { trackName: 'screen', trackSid: 0, kind: 'video' },
      { trackName: 'camera', trackSid: 1, kind: 'video' },
    ]);
    const wrapper = shallow(<ParticipantTracks participant={'mockParticipant' as any} />);
    expect(wrapper.find('Publication').length).toBe(1);
    expect(
      wrapper
        .find('Publication')
        .at(0)
        .prop('publication')
    ).toEqual({ trackName: 'camera', trackSid: 1, kind: 'video' });
  });

  describe('with enableScreenShare prop', () => {
    it('should filter out camera publications when a screen share publication is present', () => {
      mockUseVideoContext.mockImplementation(() => ({ room: {} }));
      mockUsePublications.mockImplementation(() => [
        { trackName: 'screen', trackSid: 0, kind: 'video' },
        { trackName: 'camera', trackSid: 1, kind: 'video' },
      ]);
      const wrapper = shallow(<ParticipantTracks participant={'mockParticipant' as any} enableScreenShare />);
      expect(wrapper.find('Publication').length).toBe(1);
      expect(
        wrapper
          .find('Publication')
          .at(0)
          .prop('publication')
      ).toEqual({ trackName: 'screen', trackSid: 0, kind: 'video' });
    });

    it('should render camera publications when a screen share publication is absent', () => {
      mockUseVideoContext.mockImplementation(() => ({ room: {} }));
      mockUsePublications.mockImplementation(() => [{ trackName: 'camera', trackSid: 1, kind: 'video' }]);
      const wrapper = shallow(<ParticipantTracks participant={'mockParticipant' as any} enableScreenShare />);
      expect(wrapper.find('Publication').length).toBe(1);
      expect(
        wrapper
          .find('Publication')
          .at(0)
          .prop('publication')
      ).toEqual({ trackName: 'camera', trackSid: 1, kind: 'video' });
    });
  });
});
