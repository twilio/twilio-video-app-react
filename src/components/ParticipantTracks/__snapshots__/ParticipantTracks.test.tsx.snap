// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the ParticipantTracks component should render an array of publications 1`] = `
<Fragment>
  <Publication
    isLocal={false}
    key="0"
    participant="mockParticipant"
    publication={
      Object {
        "trackSid": 0,
      }
    }
  />
  <Publication
    isLocal={false}
    key="1"
    participant="mockParticipant"
    publication={
      Object {
        "trackSid": 1,
      }
    }
  />
</Fragment>
`;
